26_Autoexposure/main.cpp

```diff
@@ -0,0 +1,874 @@
+using namespace nbl::video;
+
+
+int main()
+{
+	nbl::SIrrlichtCreationParameters deviceParams;
+	deviceParams.Bits = 24; //may have to set to 32bit for some platforms
+	deviceParams.ZBufferBits = 24; //we'd like 32bit here
+	deviceParams.DriverType = EDT_OPENGL; //! Only Well functioning driver, software renderer left for sake of 2D image drawing
+	deviceParams.WindowSize = dimension2d<uint32_t>(1280, 720);
+	deviceParams.Fullscreen = false;
+	deviceParams.Vsync = true; //! If supported by target platform
+	deviceParams.Doublebuffer = true;
+	deviceParams.Stencilbuffer = false; //! This will not even be a choice soon
+
+	auto device = createDeviceEx(deviceParams);
+	if (!device)
+		return 1; // could not create selected driver.
+
+	using LumaMeterClass = ext::LumaMeter::CLumaMeter;
+	constexpr auto MeterMode = LumaMeterClass::EMM_MEDIAN;
+	const float minLuma = 1.f/2048.f;
+	const float maxLuma = 65536.f;
+
+	auto cpuLumaMeasureSpecializedShader = LumaMeterClass::createShader(glslCompiler,inputColorSpace,MeterMode,minLuma,maxLuma);
+	auto gpuLumaMeasureShader = driver->createShader(smart_refctd_ptr<const ICPUShader>(cpuLumaMeasureSpecializedShader->getUnspecialized()));
+	auto gpuLumaMeasureSpecializedShader = driver->createSpecializedShader(gpuLumaMeasureShader.get(), cpuLumaMeasureSpecializedShader->getSpecializationInfo());
+
+	const float meteringMinUV[2] = { 0.1f,0.1f };
+	const float meteringMaxUV[2] = { 0.9f,0.9f };
+	LumaMeterClass::Uniforms_t<MeterMode> uniforms;
+	auto lumaDispatchInfo = LumaMeterClass::buildParameters(uniforms, outImg->getCreationParameters().extent, meteringMinUV, meteringMaxUV);
+
+	auto uniformBuffer = driver->createFilledDeviceLocalBufferOnDedMem(sizeof(uniforms),&uniforms);
+
+
+	using ToneMapperClass = ext::ToneMapper::CToneMapper;
+	constexpr auto TMO = ToneMapperClass::EO_ACES;
+	constexpr bool usingLumaMeter = MeterMode<LumaMeterClass::EMM_COUNT;
+	constexpr bool usingTemporalAdapatation = true;
+
+	auto cpuTonemappingSpecializedShader = ToneMapperClass::createShader(am->getGLSLCompiler(),
+		inputColorSpace,
+		std::make_tuple(outFormat,ECP_SRGB,OETF_sRGB),
+		TMO,usingLumaMeter,MeterMode,minLuma,maxLuma,usingTemporalAdapatation
+	);
+	auto gpuTonemappingShader = driver->createShader(smart_refctd_ptr<const ICPUShader>(cpuTonemappingSpecializedShader->getUnspecialized()));
+	auto gpuTonemappingSpecializedShader = driver->createSpecializedShader(gpuTonemappingShader.get(),cpuTonemappingSpecializedShader->getSpecializationInfo());
+
+	auto outImgStorage = ToneMapperClass::createViewForImage(driver,false,core::smart_refctd_ptr(outImg),{static_cast<IImage::E_ASPECT_FLAGS>(0u),0,1,0,1});
+
+	auto parameterBuffer = driver->createDeviceLocalGPUBufferOnDedMem(ToneMapperClass::getParameterBufferSize<TMO,MeterMode>());
+	constexpr float Exposure = 0.f;
+	constexpr float Key = 0.18;
+	auto params = ToneMapperClass::Params_t<TMO>(Exposure, Key, 0.85f);
+	{
+		params.setAdaptationFactorFromFrameDelta(0.f);
```

Comment:
we used this basically
https://en.wikipedia.org/wiki/Exponential_smoothing

to blend the EV value over time

Reply:
0 was a special value to "reset" the exposure (allow it to use the instantaneous value)